lib/message-handler.js

```diff
@@ -184,532 +190,538 @@ class MessageHandler {
 
     // Monster method for inserting new messages to a mailbox
     // TODO: Refactor into smaller pieces
-    add(options, callback) {
+    async addAsync(options) {
         if (!options.prepared && options.raw && options.raw.length > consts.MAX_ALLOWED_MESSAGE_SIZE) {
-            return setImmediate(() => callback(new Error('Message size ' + options.raw.length + ' bytes is too large')));
+            throw new Error('Message size ' + options.raw.length + ' bytes is too large');
         }
 
-        this.getMailbox(options, (err, mailboxData) => {
-            if (err) {
-                return callback(err);
-            }
+        // get target mailbox data
+        // get target user data
+        let [mailboxData, userData] = await Promise.all([this.getMailboxAsync(options), this.users.collection('users').findOne({ _id: options.user })]);
```

Comment:
Should verify if userData exists after this operation. getMailboxAsync throws if mailbox is missing but findOne does not